From 91c45954776a4e92fb4c73bb7e83ec17de033680 Mon Sep 17 00:00:00 2001
From: Varsha Maira <vmaira@habana.ai>
Date: Mon, 05 Aug 2024 10:38:22 +0300
Subject: [PATCH] [SW-195014] gaudi3: tpm: fix manifest hash calc in ppboot

Earlier, the manifest image used for calculating the manifest
hash included the public key. During, signing process the
public key gets modified. Because of this manifest image used
to generate the gaudi3-pcr.bin and the signed manifest image
sent to card during FW update are different. This causes the
TPM PCR mismatch if image signing is used.

This patch removes the public key area from the image used for
manifest hash calculation in ppboot.

Change-Id: I23b4a2761484a928f3b734a28518079ba34307bd
Signed-off-by: Varsha Maira <vmaira@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index ed7d4ce..bae5d65 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -1465,17 +1465,18 @@
 
 void ppboot_compute_img_ptr_hash(u8 *hash)
 {
-	u32 img_ptrs_size, img_ptrs_sig_size, img_plus_key_len;
+	u32 img_ptrs_size, img_ptrs_img_size, img_ptrs_pub_key_size, img_ptrs_sig_size;
 	u8 crc_bytes, num_crc;
 
 	memcpy(&img_ptrs_size, (void *)(uintptr_t)IMG_PTRS_SIZE_OFFSET, 4);
 	memcpy(&img_ptrs_sig_size, (void *)(uintptr_t)IMG_PTRS_SIG_SIZE_OFFSET, 4);
+	memcpy(&img_ptrs_pub_key_size, (void *)(uintptr_t)IMG_PTRS_PUB_KEY_SIZE_OFFSET, 4);
 
 	num_crc = ppboot_get_num_crc(img_ptrs_size);
 	crc_bytes = num_crc * CRC32_SIZE;
-	img_plus_key_len = img_ptrs_size - img_ptrs_sig_size - crc_bytes;
+	img_ptrs_img_size = img_ptrs_size - img_ptrs_pub_key_size - img_ptrs_sig_size - crc_bytes;
 
-	sha384_csum((void *)(uintptr_t)IMG_PTRS_OFFSET, img_plus_key_len, hash, 0);
+	sha384_csum((void *)(uintptr_t)IMG_PTRS_OFFSET, img_ptrs_img_size, hash, 0);
 }
 
 bool get_img_ptr_hash(u8 *hash)
diff --git a/arch/gaudi3/arc/common.h b/arch/gaudi3/arc/common.h
index 5b4878b..6b02292 100644
--- a/arch/gaudi3/arc/common.h
+++ b/arch/gaudi3/arc/common.h
@@ -122,6 +122,8 @@
 					offsetof(struct hl_fw_boot_ptrs, boot_ptrs_img_sz))
 #define IMG_PTRS_SIG_SIZE_OFFSET	(IMG_PTRS_OFFSET + \
 					offsetof(struct hl_fw_boot_ptrs, sig_sz))
+#define IMG_PTRS_PUB_KEY_SIZE_OFFSET	(IMG_PTRS_OFFSET + \
+					offsetof(struct hl_fw_boot_ptrs, pub_key_sz))
 
 #define SW_STATUS_SPI_BOOT_DESC_DMA_START	4
 #define SW_STATUS_SPI_BOOT_DESC_DMA_START_MASK	BIT(SW_STATUS_SPI_BOOT_DESC_DMA_START)
